cmake/helpers.cmake

```diff
@@ -59,17 +59,32 @@ endmacro()
 
 function(add_ur_target_compile_options name)
     if(NOT MSVC)
+        target_compile_definitions(${name} PRIVATE -D_FORTIFY_SOURCE=2)
         target_compile_options(${name} PRIVATE
             -fPIC
             -Wall
             -Wpedantic
             -Wempty-body
-            -Wunused-parameter
             $<$<CXX_COMPILER_ID:GNU>:-fdiagnostics-color=always>
             $<$<CXX_COMPILER_ID:Clang,AppleClang>:-fcolor-diagnostics>
         )
+        if (UR_HARDEN)
+            target_compile_options(${name} PRIVATE
+                -fstack-protector
+                $<$<CXX_COMPILER_ID:GNU,Clang>:-fstack-clash-protection> # Not supported in apple clang
+                -fcf-protection=full
+                -fstack-protector-strong
+                -fvisibility=hidden # Required for -fsanitize=cfi
+                # -fsanitize=cfi requires -flto, which breaks a lot of things
+                # See: https://github.com/oneapi-src/unified-runtime/issues/2120
+                # -flto
+                # $<$<CXX_COMPILER_ID:Clang,AppleClang>:-fsanitize=cfi>
+            )
+            if (CMAKE_BUILD_TYPE STREQUAL "Release")
+                target_compile_options(${name} PRIVATE -Werror)
```

Comment:
is this just for convenience if you're hacking around in debug? I'd vote for enabling it all the time if we can